src/mtev_websocket_client.c

```diff
@@ -0,0 +1,461 @@
+}
+
+static mtev_boolean
+mtev_websocket_client_send_handshake(mtev_websocket_client_t *client) {
+  char reqheader[4096];
+
+  mtev_websocket_client_create_key(client->client_key);
+
+  int reqlen = snprintf(reqheader, sizeof(reqheader),
+                        "GET %s HTTP/1.1\r\n"
+                        "Host: %s\r\n"
+                        "Upgrade: websocket\r\n"
+                        "Connection: Upgrade\r\n"
+                        "Sec-WebSocket-Key: %s\r\n"
+                        "Sec-WebSocket-Protocol: %s\r\n"
+                        "Sec-WebSocket-Version: 13\r\n"
+                        "\r\n",
+                        client->path, client->host, client->client_key, client->service);
+
+  return send_reqheader(client->e, reqheader, reqlen, &client->wanted_eventer_mask);
+}
+
+static mtev_boolean
+mtev_websocket_client_recv_handshake(mtev_websocket_client_t *client) {
+  char resheader[8192];
+  char accept_key[mtev_b64_encode_len(SHA_DIGEST_LENGTH) + 1];
+
+  mtev_http_create_websocket_accept_key(accept_key,
+                                        mtev_b64_encode_len(SHA_DIGEST_LENGTH) + 1,
+                                        client->client_key);
+
+  if(recv_resheader(client->e, resheader, sizeof(resheader), &client->wanted_eventer_mask) == -1) {
+    return mtev_false;
+  }
+
+  char *res_accept_key = strstr(resheader, "Sec-WebSocket-Accept");
+  if(res_accept_key == NULL) {
+    mtevL(mtev_error, "Websocket client couldn't find accept key in response headers\n");
+    return mtev_false;
+  }
+  /* skip the header text */
+  res_accept_key += 22; 
+
+  if(!strncmp(accept_key, res_accept_key, mtev_b64_encode_len(SHA_DIGEST_LENGTH))) {
+    return mtev_true;
+  } else {
+    mtevL(mtev_error, "Websocket client found incorrect accept key in response headers\n");
+    return mtev_false;
+  }
+}
+
+static void
+mtev_websocket_client_cleanup(mtev_websocket_client_t *client);
+
+static int
+mtev_websocket_client_drive(eventer_t e, int mask, void *closure, struct timeval *now) {
+  mtev_websocket_client_t *client = closure;
+
+  if(mask & EVENTER_EXCEPTION || client->should_close || client->closed) {
+abort_drive:
+    mtev_websocket_client_cleanup(client);
+    return 0;
+  }
+
+  if(!client->did_handshake) {
+    if(!client->sent_handshake && mask & EVENTER_WRITE) {
+      if(mtev_websocket_client_send_handshake(client) == mtev_false) {
+        mtevL(mtev_error, "mtev_websocket_client_send_handshake failed, aborting drive\n");
+        goto abort_drive;
+      }
+      client->sent_handshake = mtev_true;
+    }
+    
+    if(client->sent_handshake && mask & EVENTER_READ) {
+      if(mtev_websocket_client_recv_handshake(client) == mtev_false) {
+        mtevL(mtev_error, "mtev_websocket_client_recv_handshake failed, aborting drive\n");
+        goto abort_drive;
+      }
+      wslay_event_context_client_init(&client->wslay_ctx, &wslay_callbacks, client);
+      client->did_handshake = mtev_true;
```

Comment:
I would add a client call back here (and in the case) so the programmer can react to the completion "ready state" of the websockets connection.